src/passes/OptimizeCasts.cpp

```diff
@@ -102,6 +120,256 @@ namespace wasm {
+      // select the first RefAs we come accross which is moveable.
+      refAsReplacement.emplace(moveableExpressions.front(), curr);
+      if (fallthrough == moveableExpressions.front()) {
+        redundantRefAsReplacement.insert(fallthrough);
+      }
+    }
+  }
+
+  void visitRefCast(RefCast* curr, LocalGet* fallthrough) {
+    auto iter = refCastReplacement.find(moveableExpressions.front());
+    if (iter == refCastReplacement.end()) {
+      refCastReplacement.emplace(moveableExpressions.front(), curr);
+      if (fallthrough == moveableExpressions.front()) {
+        redundantRefCastReplacement.insert(fallthrough);
+      }
+    } else if (curr->type != iter->second->type &&
+               Type::isSubType(curr->type, iter->second->type)) {
+      iter->second = curr;
+      if (fallthrough != iter->first) {
+        redundantRefCastReplacement.erase(iter->first);
+      }
+    }
+  }
+
+  void removeRedundantReplacements() {
+    for (auto& redundantRefAs : redundantRefAsReplacement) {
+      refAsReplacement.erase(redundantRefAs);
+    }
+    for (auto& redundantRefCast : redundantRefCastReplacement) {
+      refCastReplacement.erase(redundantRefCast);
+    }
+  }
+};
+
+// Find the "best" cast to move earlier to another local.gets
+struct EarlyCastFinder
+  : public LinearExecutionWalker<EarlyCastFinder,
+                                 UnifiedExpressionVisitor<EarlyCastFinder>> {
+  PassOptions options;
+
+  // For each local index, we have an searcher to move casts for it.
+  std::unordered_map<Index, IndexEarlyCastSearcher> castMoveSearchers;
+
+  // Maps local.gets to the "best" RefAs to move to it.
+  std::unordered_map<Expression*, RefAs*> refAsToApply;
+
+  // Maps local.gets to the "best" RefCast to move to it.
+  std::unordered_map<Expression*, RefCast*> refCastToApply;
+
+  static void doNoteNonLinear(EarlyCastFinder* self, Expression**) {
+    self->castMoveSearchers.clear();
+  }
+
+  void visitExpression(Expression* curr) {
+    for (auto& indexSearcher : castMoveSearchers) {
+      indexSearcher.second.visitExpression(curr);
+    }
+  }
+
+  void visitLocalSet(LocalSet* curr) {
+    for (auto& indexSearcher : castMoveSearchers) {
+      indexSearcher.second.visitExpression(curr);
+    }
+    auto iter = castMoveSearchers.find(curr->index);
+    if (iter != castMoveSearchers.end()) {
+      iter->second.visitLocalSet(curr);
+    }
+  }
+
+  void visitLocalGet(LocalGet* curr) {
+    for (auto& indexSearcher : castMoveSearchers) {
+      indexSearcher.second.visitExpression(curr);
+    }
+    auto iter = castMoveSearchers.find(curr->index);
+    if (iter == castMoveSearchers.end()) {
+      RefCast testRefCast(getModule()->allocator);
+      EffectAnalyzer efa(options, *getModule());
+      efa.visit(&testRefCast);
+      IndexEarlyCastSearcher searcher(options, *getModule());
+      searcher.visitLocalGet(curr, *getModule());
+      castMoveSearchers.emplace(curr->index, searcher);
```

Comment:
IIUC the idea here is to create a Searcher for each local get index we see? I am a little worried about efficiency here, as LocalGets are among the most common instructions, and most of the time this optimization would not find work to do. And the Searcher class above looks like it has a significant size, so creating it is not very cheap.

This might be ok as it is. But perhaps there is a more efficient way? I am thinking, vaguely, of something like this:

* When we see a LocalGet, we just note the existence of the get for that index, perhaps adding to a `std::vector<SmallVector<Info>>` in which `vector[index]` is the vector of info for each get (which is typically small, so a SmallVector is more efficient). The info might be the get + the type it is cast to if it has a parent that is a cast - I hope we don't need any more.
* When we see a trap effect, we clear that vector, since we can't move past a trap. That might hard-code the fact that casts have only that effect, but I think it's ok if the efficiency gain is significant, and if we have an assertion elsewhere that verifies that is the only effect.
* When we see a cast of a get, we look in that vector and see if we can improve things, and if so we make a note for later.

I haven't thought that through entirely, but what do you think of the direction?